storage/postgres/notificator.go

```diff
@@ -257,23 +267,53 @@ type notifyEventPayload struct {
 	Revision       int64  `json:"revision"`
 }
 
-func (n *Notificator) processNotifications(notificationChannel <-chan *pq.Notification) {
+func (n *Notificator) processNotifications(notificationChannel <-chan *pq.Notification, processingContext context.Context) {
 	defer func() {
+		n.connectionMutex.Lock()
+		defer n.connectionMutex.Unlock()
+		n.stopProcessing() // closing context if not already closed
+		n.closeAllConsumers()
+		log.C(n.ctx).Debugf("Stop listening notification channel %s", postgresChannel)
 		atomic.StoreInt32(&n.isListening, aFalse)
-	}()
-	for pqNotification := range notificationChannel {
-		if pqNotification == nil {
-			continue
+		if atomic.LoadInt32(&n.isConnected) == aTrue {
+			if err := n.connection.Unlisten(postgresChannel); err != nil {
+				log.C(n.ctx).WithError(err).Errorf("Could not unlisten channel %s", postgresChannel)
+			}
 		}
-		payload, err := getPayload(pqNotification.Extra)
-		if err != nil {
-			log.C(n.ctx).WithError(err).Error("could not unmarshal notification payload")
-			n.closeAllConsumers() // Ensures no notifications are lost
-		} else {
-			if err = n.processNotificationPayload(payload); err != nil {
-				log.C(n.ctx).WithError(err).Error("closing consumers")
-				n.closeAllConsumers() // Ensures no notifications are lost
+	}()
+	lastNotificationReceived := time.Now()
+	for {
+		select {
+		case pqNotification, ok := <-notificationChannel:
+			if !ok {
+				log.C(n.ctx).Error("Notification channel closed")
+				return
+			}
+			if pqNotification == nil { // when connection is re-established a nil notification is sent by the library
+				log.C(n.ctx).Debug("Empty notification received")
+				continue
+			}
+			lastNotificationReceived = time.Now()
+			log.C(n.ctx).Debugf("Received new notification from channel %s", pqNotification.Channel)
+			payload, err := getPayload(pqNotification.Extra)
+			if err != nil {
+				log.C(n.ctx).WithError(err).Error("Could not unmarshal notification payload")
+				return
+			} else {
+				if err = n.processNotificationPayload(payload); err != nil {
+					log.C(n.ctx).WithError(err).Error("Closing consumers")
```

Comment:
the error could be `could not process notification payload. Closing consumers...`